### PR TITLE
PT-6671: Generate X509 certificate

### PR DESCRIPTION
**Pull request by akak1977**
## Description
Remove from virto platform built-in certificate that is being used for JWT token signing and replace it with new automatically generated self-signed certificate version that can be stored and shared across multiple platform instances.
### Jira-link:
https://virtocommerce.atlassian.net/browse/PT-6463
https://virtocommerce.atlassian.net/browse/PT-6671

### CODE REVIEW

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -54,7 +59,59 @@ public static IServiceCollection AddSecurityServices(this IServiceCollection ser
...
             return services;
         }
+
+        /// <summary>
+        /// Get currently installed certificate from the storage
+        /// or create new self-signed
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <returns></returns>
+        public static ServerCertificate TryLoadCertificate(this IConfiguration configuration)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:73]: Move to ConfigurationExtensions.cs

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -54,7 +59,59 @@ public static IServiceCollection AddSecurityServices(this IServiceCollection ser
...
+
+        /// <summary>
+        /// Save newly generated (if it so) server certificate to the shared storage.
+        /// The call of this method should be under distributed lock between platform instances
+        /// to synchronize old certs checks
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="currentCert"></param>
+        /// <exception cref="PlatformException"></exception>
+        public static void UpdateServerCertificateIfNeed(this IApplicationBuilder app, ServerCertificate currentCert)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:93]: Move to ApplicationBuilderExtensions.cs

**Review by tatarincev — DISMISSED** [review]

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:74]: Moved to repository

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs (hunk 2 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:94]: Moved to repository

**Review by ksavosteev — APPROVED** [review]

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -54,7 +60,58 @@ public static IServiceCollection AddSecurityServices(this IServiceCollection ser
...
+                // But we should check if old cert is stored
+                if (possiblyOldCert.StoredInDb && !possiblyOldCert.Expired)
+                {
+                    throw new ServerCertificateReplacedException("The other starting up instance of the platform replaced server certificate. Current instance should be restarted to use new server certificate.");
+                }
+
+                if (!string.IsNullOrEmpty(possiblyOldCert.Id))
+                {
+                    // Delete old certificate in case of one exists (the Id of newly generated certs is null)
+                    certificateService.DeleteAsync(new string[] { possiblyOldCert.Id });
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:110]: Awaiting missed.

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -54,7 +60,58 @@ public static IServiceCollection AddSecurityServices(this IServiceCollection ser
...
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="currentCert"></param>
+        /// <exception cref="PlatformException"></exception>
+        public static void UpdateServerCertificateIfNeed(this IApplicationBuilder app, ServerCertificate currentCert)
+        {
+            var configuration = app.ApplicationServices.GetService<IConfiguration>();
+            var certificateService = app.ApplicationServices.GetService<ICrudService<ServerCertificate>>();
+            var possiblyOldCert = SecurityRepository.LoadCurrentlySetServerCertificate(configuration); // Previously stored cert (possibly old, default or just created by another platform instance)
+            if (!possiblyOldCert.SerialNumber.EqualsInvariant(currentCert.SerialNumber))
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:99]: Invert this if, please.

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -54,7 +60,58 @@ public static IServiceCollection AddSecurityServices(this IServiceCollection ser
...
             return services;
         }
+
+        /// <summary>
+        /// Get currently installed certificate from the storage
+        /// or create new self-signed
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <returns></returns>
+        public static ServerCertificate TryLoadCertificate(this IConfiguration configuration)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:74]: Rename to `GetCertificate`, please.

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            get
+            {
+                return _PublicCertBytes;
+            }
+            set
+            {
+                _PublicCertBytes = value;
+                X509Certificate = new X509Certificate2(_PublicCertBytes);
+                SerialNumber = X509Certificate.SerialNumber;
+            }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:22]: Move all code from the setter to the constructor.

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+        }
+        public byte[] PrivateKeyCertBytes { get; set; }
+        public string PrivateKeyCertPassword { get; set; }
+        public bool StoredInDb { get; set; }
+        public string SerialNumber { get; private set; } = SerialNumberOfVirtoPredefined;
+        public X509Certificate2 X509Certificate { get; private set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:29]: Make this private field.

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Security.Cryptography.X509Certificates;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.Platform.Core.Security
+{
+    public class ServerCertificate : Entity, ICloneable
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:7]: Make sealed.

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Security.Cryptography.X509Certificates;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.Platform.Core.Security
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:5]:
```suggestion
namespace VirtoCommerce.Platform.Core.Security;
```

**Review by Vectorfield4 — DISMISSED** [review]

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs (hunk 9 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:5]: Aggree

File: src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs (hunk 8 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Core/Security/ServerCertificate.cs:7]: Aggree

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs (hunk 5 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:74]: GetServerCertificate

File: src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs (hunk 3 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Security/ServiceCollectionExtensions.cs:110]: Aggree

**Pull request merged**